File: 4_add_two_numbers/main.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        result = ListNode(0)
        elem = result
        carry = 0
        
        while l1 or l2 or carry:
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            
            dig_sum = val1 + val2 + carry
            carry = dig_sum // 10
            elem.next = ListNode(dig_sum % 10)
            elem = elem.next
            
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next
        
        return result.next

if __name__ == "__main__":
    l1 = ListNode(2, ListNode(4, ListNode(3)))  
    l2 = ListNode(5, ListNode(6, ListNode(4)))  
    
    solution = Solution()
    result = solution.addTwoNumbers(l1, l2)
    
    while result:
        print(result.val, end="")
        result = result.next
    print()  

File: 4_add_two_numbers/Main.java

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode result = new ListNode();
        ListNode current = result;
        int carry = 0;

        while (l1 != null || l2 != null || carry > 0) {
            int sum = carry;
            if (l1 != null) {
                sum += l1.val;
                l1 = l1.next;
            }
            if (l2 != null) {
                sum += l2.val;
                l2 = l2.next;
            }
            carry = sum / 10;
            current.next = new ListNode(sum % 10);
            current = current.next;
        }

        return result.next;
    }
}

public class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        // Example input: 2 -> 4 -> 3 and 5 -> 6 -> 4
        ListNode l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
        ListNode l2 = new ListNode(5, new ListNode(6, new ListNode(4)));
        
        ListNode result = solution.addTwoNumbers(l1, l2);
        
        // Print the result
        while (result != null) {
            System.out.print(result.val + " ");
            result = result.next;
        }
    }
}

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }

    public void append(int val) {
        ListNode current = this;
        while (current.next != null) {
            current = current.next;
        }
        current.next = new ListNode(val);
    }
}